mitodl/ocw-studio
# Concourse pipelines, take 2

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #368
Closes #384

#### What's this PR do?
- Reverts a previous revert
- Creates an abstract and concourse-specific class for creating publishing pipelines
- If a specific pipeline class is set, instantiates it and runs concourse-ci API requests to generate draft/live pipelines for a Website when it is created via the ocw-studio API.

#### How should this be manually tested?
Kind of tricky, but this should work:

- Unpause and run the `ocw/draft/ocw-www-test` pipeline on the RC instance of concourse, which is deliberately misconfigured to clear out nearly everything on the draft S3 bucket.  After a minute or so verify that the draft site url brings up a 'NoSuchKey' error.  Pause the pipeline again.


- Create a `Website` with name "ocw-www" based on the "ocw-www" starter config from github.com/mitodl/ocw-hugo-projects.  The starter should a slug of "ocw-www" and `"root-url-path": ""` in its config.
- Create a `Website` with name "course-mb-1" based on the "ocw-course" starter config from github.com/mitodl/ocw-hugo-projects. The starter should a slug of "ocw-course" and `"root-url-path": "courses"` in its config.

- Set the following env variables to the same as on RC:
```
OCW_STUDIO_BASE_URL
GIT_ORGANIZATION
```

Plus set the following:
```
GIT_DOMAIN=github.mit.edu
ROOT_WEBSITE_NAME=ocw-www
CONTENT_SYNC_PIPELINE=content_sync.pipelines.concourse.ConcourseGithubPipeline
CONCOURSE_URL=<QA instance of concourse>
CONCOURSE_USERNAME=<ask me>
CONCOURSE_PASSWORD=<ask me>
AWS_PREVIEW_BUCKET_NAME=<ask me>
AWS_PUBLISH_BUCKET_NAME=<ask me>
```

To simulate creating a pipeline for the home page:
- Insert this into `content_sync.pipelines.concourse.ConcourseGithubPipeline.upsert_website_pipeline`, on the 1st line:
```python
from uuid import UUID    
self.website.uuid = UUID("eb8b7017-4c71-4752-a91d-be730af76769") # match the site in ocw-studio-rc
```

- Run the following:
```python
from websites.models import *
from content_sync.pipelines.concourse import *
website = Website.objects.get(name="ocw-www")
pipeline = ConcourseGithubPipeline(website)
pipeline.upsert_website_pipeline()
```

- Go to the QA concourse site at ../teams/ocw/pipelines/draft/jobs/build-ocw-site/builds/2?vars.site="ocw-www" and start a new build.  It should complete successfully and you should see a full home page at the draft fastly url.  Click on the testimonials and you should be brought to a working testimonial page with no missing images.



To simulate creating a pipeline for a course:
- Replace the code change above with this:
```python        
self.website.uuid = UUID("917f160c-7eae-48d8-86a6-a041bd2ef0fe"). # match uuid in ocw-studio-rc
```        
- Restart your shell and run the following:
```python
from websites.models import *
from content_sync.pipelines.concourse import *
website = Website.objects.get(name="course-mb-1")
pipeline = ConcourseGithubPipeline(website)
pipeline.upsert_website_pipeline()
```

- Go to the QA concourse site at ..teams/ocw/pipelines/draft/jobs/build-ocw-site/builds/3?vars.site="course-mb-1" and start a new build.  It should complete successfully and you should see content at the following urls on the fastly draft site:
/courses/course-mb-1/superb-owl/
/courses/course-mb-1/527fc14517b14226a30aa8f21e42a0a4_DSC02800a.jpg
